tools/libboard/board_n3000/board_n3000.c

```diff
@@ -313,25 +270,328 @@ fpga_result print_board_info(fpga_token token)
 	return res;
 }
 
+//enum eth group feature dfl_dev*
+fpga_result enum_eth_group_feature(fpga_token token,
+				   char eth_feature[ETH_GROUP_COUNT][SYSFS_MAX_SIZE],
+				   uint32_t size)
+{
+	fpga_result res          = FPGA_OK;
+	uint32_t i               = 0;
+	uint32_t index           = 0;
+	uint64_t value           = 0;
+	char sysfs_path[SYSFS_MAX_SIZE]     = { 0 };
+
+	for (i = 0; i < MAX_FPGA_FEATURE_COUNT; i++) {
+
+		value = 0;
+		if (snprintf(sysfs_path, SYSFS_MAX_SIZE,
+				"dfl_dev.%d/feature_id", i) < 0) {
+			OPAE_ERR("snprintf buffer overflow");
+			return FPGA_EXCEPTION;
+		}
+
+		res = read_sysfs_int64(token, sysfs_path, &value);
+		if (res != FPGA_OK) {
+			OPAE_MSG("Failed to read feature id");
+			continue;
+		}
+
+		if (value == ETH_GROUP_FEATURE_ID) {
+			if (snprintf(eth_feature[index], SYSFS_MAX_SIZE,
+				"dfl_dev.%d", i) < 0) {
+				OPAE_ERR("snprintf buffer overflow");
+				return FPGA_EXCEPTION;
+			}
+			index = index + 1;
+		}
+
+		if (index >= size) {
+			break;
+		}
+	}
+
+	if (index  < size) {
+		OPAE_ERR("Not found eth group Feature");
+		return FPGA_NOT_FOUND;
+	}
+
+	return res;
+}
+
+//enum pkvl regmap path
+// /sys/kernel/debug/regmap/spi%d.0/registers
+fpga_result enum_pkvl_sysfs_path(fpga_token token, char *pkvl_path)
+{
+	char sysfs_path[SYSFS_MAX_SIZE] = { 0 };
+	fpga_result res                 = FPGA_OK;
+	uint32_t i                      = 0;
+	fpga_object fpga_object;
+
+	for (i = 0; i < MAX_FPGA_FEATURE_COUNT; i++) {
+
+		if (snprintf(sysfs_path, SYSFS_MAX_SIZE,
+			"dfl*/*spi*/spi_master/spi*/spi%d*/", i) < 0) {
+			OPAE_ERR("snprintf buffer overflow");
+			return FPGA_EXCEPTION;
+		}
+
+		res = fpgaTokenGetObject(token, sysfs_path,
+						&fpga_object, FPGA_OBJECT_GLOB);
+		if (res != FPGA_OK) {
+			OPAE_MSG("Failed to get token Object");
+			continue;
+		}
+
+		res = fpgaDestroyObject(&fpga_object);
+		if (res != FPGA_OK) {
+			OPAE_ERR("Failed to Destroy Object");
+		}
```

Comment:
Is this simply confirming that the path exists?  Wasn't sure about it, because you're not fetching the value behind the object.

Reply:
1.	Find your max10 device. 
For example: /sys/class/fpga_region/region0/dfl-fme.0/dfl_dev.0/subdev_spi_altera.3.auto/spi_master/spi3/spi3.0
Then we can see the max10 device is “spi3.0”
2.	Find the max10 regmap debugfs entry.
cd /sys/kernel/debug/regmap/spi3.0
3.	Read the “registers” node.
It is not so straightforward to get the value of the register you want. The output is a very long formatted string of all registers & values in max10 register space. like:
00000000: ffffffff
00000004: ffffffff
00000008: ffffffff

and for PKVL link status, you need to set the correct reading offset and find:
00300964: XXXXXXXX
Please refer to max10 spec and double check.
4.	get the link states of each link according to max10 spec.



Hi Tim,

how  to enum/get  **spi3.0** using OPAE API's ?

Reply:
Couldn't you use the object API with your globbing extension?

Reply:
fixed